File: scatter.R
#' Create ggplot Scatter Plot
#'
#' @param dat Data frame containing `x` and `y` columns, and optional metadata variables for coloring, shaping, or faceting.
#' @param color_by Character string specifying column name to color points by, or `"none"`. Default is `"none"`.
#' @param shape_by Character string specifying column name to set point shapes by, or `"none"`. Default is `"none"`.
#' @param facet_by Character string specifying column name to facet plot by, or `"none"`. Default is `"none"`.
#' @param add_line Logical indicating whether to add regression line(s). Default is `TRUE`.
#' @param point_size Numeric specifying size of scatter points. Default is `3`.
#' @param point_alpha Numeric specifying transparency (alpha) of scatter points (0 to 1). Default is `0.7`.
#' @param x_label Character string for X-axis label. Default is `"x"`.
#' @param y_label Character string for Y-axis label. Default is `"y"`.
#'
#' @author Brian S Yandell, \email{brian.yandell@@wisc.edu}
#' @keywords utilities
#'
#' @return A `ggplot` object (or null plot object if required columns are missing).
#'
#' @export
#' @importFrom ggplot2 ggplot aes geom_point geom_smooth facet_wrap labs scale_shape_manual scale_color_brewer scale_color_hue .data
#' @importFrom shiny isTruthy
scatter_ggplot <- function(dat,
                           color_by = "none",
                           shape_by = "none",
                           facet_by = "none",
                           add_line = TRUE,
                           point_size = 3,
                           point_alpha = 0.7,
                           x_label = "x",
                           y_label = "y") {
  if (!all(c("x", "y") %in% colnames(dat))) {
    return(plot_null("dataframe must contain x and y columns"))
  }

  # Convert candidates to factors if present
  for (col in c("sex", "diet", "geno", "Genotype")) {
    if (col %in% colnames(dat)) {
      dat[[col]] <- factor(dat[[col]])
    }
  }

  # Calculate sex_diet combination
  if (all(c("sex", "diet") %in% colnames(dat))) {
    dat$sex_diet <- factor(paste(dat$sex, dat$diet, sep = "_"))
  }

  # Ensure subject/rownames exists for label
  if (!("subject" %in% colnames(dat))) {
    dat$subject <- rownames(dat)
  }

  # Set up ggplot mapping
  aes_args <- list(x = quote(x), y = quote(y), label = quote(subject))

  # Color grouping
  col_var <- color_by
  if (!is.null(col_var) && col_var != "none" && col_var %in% colnames(dat)) {
    aes_args$col <- as.name(col_var)
    aes_args$group <- as.name(col_var)
  }

  # Shape grouping
  shape_var <- shape_by
  if (!is.null(shape_var) && shape_var != "none" && shape_var %in% colnames(dat)) {
    aes_args$shape <- as.name(shape_var)
  }

  p <- ggplot2::ggplot(dat) +
    do.call(ggplot2::aes, aes_args)

  p_size <- ifelse(is.null(point_size), 3, point_size)
  p_alpha <- ifelse(is.null(point_alpha), 0.7, point_alpha)

  # Add regression lines first (so they are plotted behind/below the symbols)
  if (shiny::isTruthy(add_line)) {
    if (!is.null(col_var) && col_var != "none" && col_var %in% colnames(dat)) {
      p <- p + ggplot2::geom_smooth(ggplot2::aes(label = NULL), method = "lm", formula = y ~ x, se = FALSE, linetype = "solid", linewidth = 1)
    } else {
      p <- p + ggplot2::geom_smooth(ggplot2::aes(group = 1, label = NULL), method = "lm", formula = y ~ x, se = FALSE, col = "black", linetype = "solid", linewidth = 1)
    }
  }

  # Add scatter points on top of regression lines
  if (is.null(shape_var) || shape_var == "none") {
    p <- p + ggplot2::geom_point(shape = 1, size = p_size, alpha = p_alpha, stroke = 1.5)
  } else {
    p <- p + ggplot2::geom_point(size = p_size, alpha = p_alpha, stroke = 1.5)
    p <- p + ggplot2::scale_shape_manual(values = c(1, 2, 5, 0, 6, 3, 4, 7, 8, 9, 10, 11, 12, 13, 14))
  }

  # Faceting
  facet_var <- facet_by
  if (!is.null(facet_var) && facet_var != "none" && facet_var %in% colnames(dat)) {
    p <- p + ggplot2::facet_wrap(ggplot2::vars(.data[[facet_var]]))
  }

  # Apply theme and colors
  p <- p + scattyr_plot_theme()
  # Apply custom axis labels
  p <- p + ggplot2::labs(x = x_label, y = y_label)
  if (!is.null(col_var) && col_var != "none" && col_var %in% colnames(dat)) {
    num_levels <- length(unique(dat[[col_var]]))
    if (num_levels <= 8) {
      p <- p + ggplot2::scale_color_brewer(palette = "Dark2")
    } else {
      p <- p + ggplot2::scale_color_hue()
    }
  }

  p
}

File: scatter.py
"""
Scatter plot rendering using plotnine.
"""

import pandas as pd
import numpy as np
from plotnine import (
    ggplot, aes, geom_point, geom_smooth, facet_wrap, labs,
    theme_minimal, scale_shape_manual, scale_color_brewer, scale_color_hue
)
from typing import Optional


def scatter_plotnine(
    dat: pd.DataFrame,
    color_by: str = "none",
    shape_by: str = "none",
    facet_by: str = "none",
    add_line: bool = True,
    point_size: float = 3.0,
    point_alpha: float = 0.7,
    x_label: str = "x",
    y_label: str = "y"
) -> Optional[ggplot]:
    """
    Create a plotnine scatter plot mirroring scatter_ggplot in R.

    Parameters
    ----------
    dat : pd.DataFrame
        DataFrame containing 'x' and 'y' columns, plus metadata variables.
    color_by : str
        Column name to color points by, or "none". Default is "none".
    shape_by : str
        Column name to set point shapes by, or "none". Default is "none".
    facet_by : str
        Column name to facet plot by, or "none". Default is "none".
    add_line : bool
        Whether to add linear regression line(s). Default is True.
    point_size : float
        Scatter point size. Default is 3.0.
    point_alpha : float
        Scatter point transparency (0.0 to 1.0). Default is 0.7.
    x_label : str
        X-axis label. Default is "x".
    y_label : str
        Y-axis label. Default is "y".

    Returns
    -------
    plotnine.ggplot.ggplot or None
        A plotnine ggplot object, or None if required columns are missing.
    """
    if dat is None or not isinstance(dat, pd.DataFrame):
        return None

    if "x" not in dat.columns or "y" not in dat.columns:
        return None

    df = dat.copy()

    # Convert candidates to category if present
    for col in ["sex", "diet", "geno", "Genotype"]:
        if col in df.columns:
            df[col] = df[col].astype("category")

    # Calculate sex_diet combination if both columns exist
    if "sex" in df.columns and "diet" in df.columns:
        df["sex_diet"] = (df["sex"].astype(str) + "_" + df["diet"].astype(str)).astype("category")

    # Ensure subject column exists for labeling
    if "subject" not in df.columns:
        df["subject"] = df.index.astype(str)

    mapping = {"x": "x", "y": "y"}

    col_var = color_by if (color_by and color_by != "none" and color_by in df.columns) else None
    shape_var = shape_by if (shape_by and shape_by != "none" and shape_by in df.columns) else None
    facet_var = facet_by if (facet_by and facet_by != "none" and facet_by in df.columns) else None

    if col_var:
        mapping["color"] = col_var
        mapping["group"] = col_var

    if shape_var:
        mapping["shape"] = shape_var

    p = ggplot(df, aes(**mapping))

    # Add regression lines first (so they are plotted behind/below symbols)
    if add_line:
        if col_var:
            p += geom_smooth(method="lm", se=False, linetype="solid", size=1)
        else:
            p += geom_smooth(aes(group=1), method="lm", se=False, color="black", linetype="solid", size=1)

    # Add scatter points (open symbols / transparent fill matching R ggplot shape=1)
    if not shape_var:
        p += geom_point(shape="o", fill="none", size=point_size, alpha=point_alpha, stroke=1.5)
    else:
        p += geom_point(fill="none", size=point_size, alpha=point_alpha, stroke=1.5)
        shapes = ["o", "s", "^", "D", "v", "p", "*", "+", "x", "h", "8", "H", "d", "|", "_"]
        p += scale_shape_manual(values=shapes)

    # Faceting
    if facet_var:
        p += facet_wrap(f"~ {facet_var}")

    # Theme and custom labels
    p += theme_minimal(base_size=9)
    p += labs(x=x_label, y=y_label)

    if col_var:
        num_levels = df[col_var].nunique()
        if num_levels <= 8:
            p += scale_color_brewer(type="qual", palette="Dark2")
        else:
            p += scale_color_hue()

    return p
